packages/kbn-content-management-utils/src/saved_object_content_storage.ts

```diff
@@ -0,0 +1,396 @@
+  const { query, contentTypeId, options } = params;
+  const { included, excluded } = query.tags ?? {};
+  const hasReference: SavedObjectsFindOptions['hasReference'] = included
+    ? included.map((id) => ({
+        id,
+        type: 'tag',
+      }))
+    : undefined;
+
+  const hasNoReference: SavedObjectsFindOptions['hasNoReference'] = excluded
+    ? excluded.map((id) => ({
+        id,
+        type: 'tag',
+      }))
+    : undefined;
+
+  return {
+    type: contentTypeId,
+    search: query.text,
+    perPage: query.limit,
+    page: query.cursor ? +query.cursor : undefined,
+    defaultSearchOperator: 'AND',
+    searchFields: options?.searchFields ?? ['description', 'title'],
+    fields: options?.fields ?? ['description', 'title'],
+    hasReference,
+    hasNoReference,
+  };
+};
+
+export const createArgsToSoCreateOptionsDefault = (
+  params: SavedObjectsCreateOptions
+): SavedObjectsCreateOptions => params;
+
+export const updateArgsToSoUpdateOptionsDefault = <Types extends CMCrudTypes>(
+  params: SavedObjectsUpdateOptions<Types['Attributes']>
+): SavedObjectsUpdateOptions<Types['Attributes']> => params;
+
+export type CreateArgsToSoCreateOptions<Types extends CMCrudTypes> = (
+  params: Types['CreateOptions']
+) => SavedObjectsCreateOptions;
+
+export type SearchArgsToSOFindOptions<Types extends CMCrudTypes> = (
+  params: Types['SearchIn']
+) => SavedObjectsFindOptions;
+
+export type UpdateArgsToSoUpdateOptions<Types extends CMCrudTypes> = (
+  params: Types['UpdateOptions']
+) => SavedObjectsUpdateOptions<Types['Attributes']>;
+
+export interface SOContentStorageConstrutorParams<Types extends CMCrudTypes> {
+  savedObjectType: string;
+  cmServicesDefinition: ServicesDefinitionSet;
+  createArgsToSoCreateOptions?: CreateArgsToSoCreateOptions<Types>;
+  updateArgsToSoUpdateOptions?: UpdateArgsToSoUpdateOptions<Types>;
+  searchArgsToSOFindOptions?: SearchArgsToSOFindOptions<Types>;
+  enableMSearch?: boolean;
+}
+
+export abstract class SOContentStorage<Types extends CMCrudTypes>
+  implements
+    ContentStorage<
+      Types['Item'],
+      Types['PartialItem'],
+      MSearchConfig<Types['Item'], Types['Attributes']>
+    >
+{
+  constructor({
+    savedObjectType,
+    cmServicesDefinition,
+    createArgsToSoCreateOptions,
+    updateArgsToSoUpdateOptions,
+    searchArgsToSOFindOptions,
+    enableMSearch,
+  }: SOContentStorageConstrutorParams<Types>) {
+    this.savedObjectType = savedObjectType;
+    this.cmServicesDefinition = cmServicesDefinition;
+    this.createArgsToSoCreateOptions =
+      createArgsToSoCreateOptions || createArgsToSoCreateOptionsDefault;
+    this.updateArgsToSoUpdateOptions =
+      updateArgsToSoUpdateOptions || updateArgsToSoUpdateOptionsDefault;
+    this.searchArgsToSOFindOptions = searchArgsToSOFindOptions || searchArgsToSOFindOptionsDefault;
+
+    // todo - talk to anton about a better way to implement this
+    if (enableMSearch) {
+      this.mSearch = {
```

Comment:
@Dosant We should find a better api for this. I'd prefer to use some sort of boolean to determine whether mSearch is enabled. As best I can tell this is largely a reconfiguration of code used elsewhere